### PR TITLE
APIF-317: Detect changed response defaults.

### PR DESCRIPTION
**Pull request by mpokryva**
Extend OAD to detect changing/added/removed response defaults.

### CODE REVIEW

File: src/test/java/com/qdesrame/openapi/test/BackwardCompatibilityTest.java [hunk 1]
```diff
@@ -1,10 +1,12 @@
...
 
-import static com.qdesrame.openapi.test.TestUtils.assertOpenApiBackwardCompatible;
-import static com.qdesrame.openapi.test.TestUtils.assertOpenApiBackwardIncompatible;
-
+import com.qdesrame.openapi.diff.OpenApiCompare;
+import io.swagger.v3.oas.models.OpenAPI;
+import io.swagger.v3.oas.models.responses.ApiResponse;
 import org.junit.jupiter.api.Test;
 
+import static com.qdesrame.openapi.test.TestUtils.*;
```
Comment (codyaray) [inline src/test/java/com/qdesrame/openapi/test/BackwardCompatibilityTest.java]: static star imports may not be accepted upstream either. makes tracing where imported functions come from a bit harder (without an IDE)

File: src/test/java/com/qdesrame/openapi/test/BackwardCompatibilityTest.java [hunk 2]
```diff
@@ -47,4 +49,34 @@ public void testApiOperationChanged() {
...
+  public void testApiResponseDefaultRemoved() {
+      OpenAPI spec = OpenApiCompare.openApiParser.read(OPENAPI_DOC5);
+      OpenAPI specWithDefault = specWithDefaultResponse(OPENAPI_DOC5, new ApiResponse().description("hi"));
+      assertOpenApiBackwardIncompatible(specWithDefault, spec);
+  }
+
+  @Test
+  public void testApiResponseDefaultChanged() {
+    OpenAPI specWithDefault1 = specWithDefaultResponse(OPENAPI_DOC5, new ApiResponse().description("hi"));
+    OpenAPI specWithDefault2 = specWithDefaultResponse(OPENAPI_DOC5, new ApiResponse().description("hey"));
```
Comment (codyaray) [inline src/test/java/com/qdesrame/openapi/test/BackwardCompatibilityTest.java:146]: bug: changing the description shouldn't make these incompatible...

File: src/main/java/com/qdesrame/openapi/diff/model/ChangedApiResponse.java [hunk 3]
```diff
@@ -43,6 +44,11 @@ public ChangedApiResponse(
 
   @Override
   public DiffResult isCoreChanged() {
+    ApiResponse oldDefault = oldApiResponses.getDefault();
+    ApiResponse newDefault = newApiResponses.getDefault();
+    if (!Objects.equals(oldDefault, newDefault)) {
```
Comment (codyaray) [inline src/main/java/com/qdesrame/openapi/diff/model/ChangedApiResponse.java]: this doesn't feel quite right. the `Changed*` objects are just holder objects...  I expect one of the `*Diff` objects to the `diff` bit.

And you don't want exact equality (as mentioned in the bug above)

Comment (mpokryva) [reply src/main/java/com/qdesrame/openapi/diff/model/ChangedApiResponse.java]: https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md#response-object So should the comparison be: Equality of all subfields, except description?

Comment (codyaray) [reply src/main/java/com/qdesrame/openapi/diff/model/ChangedApiResponse.java]: i don't know that its strict equality of all subfields (e.g., order of headers doesn't matter). I'm also not sure how `links` should be handled yet... these don't appear in the actual request/response body, but are annotations used for openapi processing itself.

I think somehow this needs to be used to compare the responses: https://github.com/confluentinc/openapi-diff/blob/master/src/main/java/com/qdesrame/openapi/diff/compare/ResponseDiff.java#L43

**Comment by mpokryva** [comment]
Somehow didn't realize that OAD already does this... Don't know how that happened. Anyways, reverted my changes and wrote tests for this.

File: src/test/java/com/qdesrame/openapi/test/BackwardCompatibilityTest.java [hunk 4]
```diff
@@ -54,31 +58,33 @@ public void testApiReadWriteOnlyPropertiesChanged() {
...
-      assertOpenApiBackwardIncompatible(specWithDefault, spec);
+    OpenAPI spec = OpenApiCompare.getOpenApiParser().read(OPENAPI_DOC5);
+    OpenAPI specWithDefault = specWithDefaultResponse(OPENAPI_DOC5, new ApiResponse().description("hi"));
+    assertOpenApiBackwardIncompatible(specWithDefault, spec);
   }
 
   @Test
   public void testApiResponseDefaultChanged() {
-    OpenAPI specWithDefault1 = specWithDefaultResponse(OPENAPI_DOC5, new ApiResponse().description("hi"));
-    OpenAPI specWithDefault2 = specWithDefaultResponse(OPENAPI_DOC5, new ApiResponse().description("hey"));
```
Comment (codyaray) [inline src/test/java/com/qdesrame/openapi/test/BackwardCompatibilityTest.java]: so just changing the description is "compatible" now, right?

File: src/test/java/com/qdesrame/openapi/test/TestUtils.java [hunk 5]
```diff
@@ -34,6 +34,13 @@ public static void assertOpenApiBackwardCompatible(
     assertThat(changedOpenApi.isCompatible()).isTrue();
   }
 
+  public static void assertOpenApiBackwardCompatible(
+          OpenAPI oldSpec, OpenAPI newSpec, boolean isDiff) {
```
Comment (codyaray) [inline src/test/java/com/qdesrame/openapi/test/TestUtils.java]: nit: `isDiff` is unused

**Review by codyaray — APPROVED** [review]

File: src/test/java/com/qdesrame/openapi/test/BackwardCompatibilityTest.java (hunk 4 above)
Comment (mpokryva) [reply src/test/java/com/qdesrame/openapi/test/BackwardCompatibilityTest.java]: Yep!

File: src/test/java/com/qdesrame/openapi/test/TestUtils.java (hunk 5 above)
Comment (mpokryva) [reply src/test/java/com/qdesrame/openapi/test/TestUtils.java]: Used now!

**Pull request merged**